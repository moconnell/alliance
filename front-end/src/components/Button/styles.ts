import styled from "styled-components";

export const Container = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  background: #ff3864;
  border-radius: 6px;
`;

export const ButtonComponent = styled.button`
  padding: 8px 16px;
  color: #fff;
  font-family: Inter;
  font-size: 14px;
  font-weight: 700;
  line-height: 20px;
  width: 100%;

  :hover {
    opacity: 0.7;
  }

  ${({ disabled }: any) =>
    disabled &&
    `
      pointer-events: none;
      opacity: 0.7;

      :hover{
      opacity: 0.7;
      }
    `}
`;
