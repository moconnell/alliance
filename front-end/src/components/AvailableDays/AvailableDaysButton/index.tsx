import "./styles.css";
import { useEffect, useState } from "react";
import DaysOfWeek from "../../../types/daysOfWeek";

interface AvailableDaysButtonProps {
  day: DaysOfWeek;
  selected?: boolean;
  onSelect?: (day: DaysOfWeek, selected: boolean) => void;
}

const AvailableDaysButton = ({
  day,
  selected = false,
  onSelect,
}: AvailableDaysButtonProps) => {
  const [active, setActive] = useState(selected);

  useEffect(() => {
    setActive(selected);
  }, [selected]);

  const handleClick = () => {
    const newActive = !active;
    setActive(newActive);
    if (onSelect) onSelect(day, newActive);
  };

  return (
    <div className="circle-wrapper" role="button" aria-pressed={active} onClick={handleClick}>
      <div className={active ? "name-wrapper-clicked" : "name-wrapper"}>
        <span className={active ? "clicked-dot" : "dot"}></span>
        {DaysOfWeek[day].substring(0, 3)}
      </div>
    </div>
  );
};

export default AvailableDaysButton;
