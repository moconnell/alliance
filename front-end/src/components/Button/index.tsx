import React from "react";
import { Container, ButtonComponent } from "./styles";

interface ButtonProps {
  children?: React.ReactNode;
  title?: string;
  disabled?: boolean;
  testId?: string;
  type?: "button" | "submit" | "reset";
  onClick?: React.MouseEventHandler;
}

const Button = ({
  children,
  disabled,
  title,  
  testId = "button",
  type = "button",
  onClick,
}: ButtonProps) => {
  return (
    <Container aria-disabled={disabled}>
      <ButtonComponent data-testid={testId} disabled={disabled} title={title} onClick={onClick} type={type}>
        {children}
      </ButtonComponent>
    </Container>
  );
};

export default Button;
