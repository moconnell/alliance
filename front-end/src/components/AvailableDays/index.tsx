import { Flex } from "@chakra-ui/react";
import React, { useEffect, useState } from "react";
import DaysOfWeek from "../../types/daysOfWeek";
import AvailableDaysButton from "./AvailableDaysButton";

const orderedDaysOfWeek = [
  DaysOfWeek.Monday,
  DaysOfWeek.Tuesday,
  DaysOfWeek.Wednesday,
  DaysOfWeek.Thursday,
  DaysOfWeek.Friday,
  DaysOfWeek.Saturday,
  DaysOfWeek.Sunday,
];
interface AvailableDaysProps {
  value?: DaysOfWeek;
  onChange?: (value: DaysOfWeek) => void;
}

const AvailableDays = ({
  value = DaysOfWeek.None,
  onChange,
}: AvailableDaysProps) => {
  const [availability, setAvailability] = useState(value);

  useEffect(() => {
    setAvailability(value);
  }, [value]);

  const handleSelect = (day: DaysOfWeek) => {
    const newAvailability = availability ^ day;
    setAvailability(newAvailability);
    if (onChange) onChange(newAvailability);
  };

  return (
    <Flex direction={{ base: "row" }}>
      {orderedDaysOfWeek.map((day) => (
        <AvailableDaysButton
          day={day}
          key={day}
          selected={(day & availability) === day}
          onSelect={handleSelect}
        />
      ))}
    </Flex>
  );
};

export default AvailableDays;
