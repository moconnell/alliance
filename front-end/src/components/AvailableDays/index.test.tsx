import { fireEvent, render } from "@testing-library/react";
import { useState } from "react";
import DaysOfWeek from "../../types/daysOfWeek";
import Button from "../Button";
import AvailableDays from "./index";

const TestComponent = () => {
  const [value, setValue] = useState(DaysOfWeek.MonFri);

  return (
    <>
      <AvailableDays value={value} />
      <Button testId="button:update" onClick={() => setValue(DaysOfWeek.Monday)} />
    </>
  );
};

describe("<AvailableDays />", () => {
  it("should render correctly", async () => {
    const { asFragment } = render(<AvailableDays />);
    expect(asFragment()).toMatchSnapshot();
  });

  const testData = [
    {
      defaultValue: DaysOfWeek.Monday,
      clickText: /Tue/i,
      expectedValue: DaysOfWeek.Monday | DaysOfWeek.Tuesday,
    },
    {
      defaultValue: DaysOfWeek.MonFri,
      clickText: /Fri/i,
      expectedValue:
        DaysOfWeek.Monday |
        DaysOfWeek.Tuesday |
        DaysOfWeek.Wednesday |
        DaysOfWeek.Thursday,
    },
  ];

  testData.forEach(({ defaultValue, clickText, expectedValue }) =>
    it("should send value onChange", async () => {
      const mockHandleChange = jest.fn();
      const { getByText } = render(
        <AvailableDays value={defaultValue} onChange={mockHandleChange} />
      );
      fireEvent.click(getByText(clickText));
      expect(mockHandleChange).toHaveBeenCalledWith(expectedValue);
    })
  );

  it("should re-render after update", async () => {
    const { asFragment, getByTestId, getByText } = render(<TestComponent />);
    expect(asFragment()).toMatchSnapshot();
    
    fireEvent.click(getByText("Fri"));
    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(getByTestId("button:update"));
    expect(asFragment()).toMatchSnapshot();
  });
});
