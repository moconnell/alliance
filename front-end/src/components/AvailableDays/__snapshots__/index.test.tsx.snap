// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AvailableDays /> should re-render after update 1`] = `
<DocumentFragment>
  <div
    class="css-motgx2"
  >
    <div
      aria-pressed="true"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper-clicked"
      >
        <span
          class="clicked-dot"
        />
        Mon
      </div>
    </div>
    <div
      aria-pressed="true"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper-clicked"
      >
        <span
          class="clicked-dot"
        />
        Tue
      </div>
    </div>
    <div
      aria-pressed="true"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper-clicked"
      >
        <span
          class="clicked-dot"
        />
        Wed
      </div>
    </div>
    <div
      aria-pressed="true"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper-clicked"
      >
        <span
          class="clicked-dot"
        />
        Thu
      </div>
    </div>
    <div
      aria-pressed="true"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper-clicked"
      >
        <span
          class="clicked-dot"
        />
        Fri
      </div>
    </div>
    <div
      aria-pressed="false"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper"
      >
        <span
          class="dot"
        />
        Sat
      </div>
    </div>
    <div
      aria-pressed="false"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper"
      >
        <span
          class="dot"
        />
        Sun
      </div>
    </div>
  </div>
  <div
    class="sc-bczRLJ bbkjqh"
  >
    <button
      class="sc-gsnTZi kKvmkM"
      data-testid="button:update"
      type="button"
    />
  </div>
</DocumentFragment>
`;

exports[`<AvailableDays /> should re-render after update 2`] = `
<DocumentFragment>
  <div
    class="css-motgx2"
  >
    <div
      aria-pressed="true"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper-clicked"
      >
        <span
          class="clicked-dot"
        />
        Mon
      </div>
    </div>
    <div
      aria-pressed="true"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper-clicked"
      >
        <span
          class="clicked-dot"
        />
        Tue
      </div>
    </div>
    <div
      aria-pressed="true"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper-clicked"
      >
        <span
          class="clicked-dot"
        />
        Wed
      </div>
    </div>
    <div
      aria-pressed="true"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper-clicked"
      >
        <span
          class="clicked-dot"
        />
        Thu
      </div>
    </div>
    <div
      aria-pressed="false"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper"
      >
        <span
          class="dot"
        />
        Fri
      </div>
    </div>
    <div
      aria-pressed="false"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper"
      >
        <span
          class="dot"
        />
        Sat
      </div>
    </div>
    <div
      aria-pressed="false"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper"
      >
        <span
          class="dot"
        />
        Sun
      </div>
    </div>
  </div>
  <div
    class="sc-bczRLJ bbkjqh"
  >
    <button
      class="sc-gsnTZi kKvmkM"
      data-testid="button:update"
      type="button"
    />
  </div>
</DocumentFragment>
`;

exports[`<AvailableDays /> should re-render after update 3`] = `
<DocumentFragment>
  <div
    class="css-motgx2"
  >
    <div
      aria-pressed="true"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper-clicked"
      >
        <span
          class="clicked-dot"
        />
        Mon
      </div>
    </div>
    <div
      aria-pressed="false"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper"
      >
        <span
          class="dot"
        />
        Tue
      </div>
    </div>
    <div
      aria-pressed="false"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper"
      >
        <span
          class="dot"
        />
        Wed
      </div>
    </div>
    <div
      aria-pressed="false"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper"
      >
        <span
          class="dot"
        />
        Thu
      </div>
    </div>
    <div
      aria-pressed="false"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper"
      >
        <span
          class="dot"
        />
        Fri
      </div>
    </div>
    <div
      aria-pressed="false"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper"
      >
        <span
          class="dot"
        />
        Sat
      </div>
    </div>
    <div
      aria-pressed="false"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper"
      >
        <span
          class="dot"
        />
        Sun
      </div>
    </div>
  </div>
  <div
    class="sc-bczRLJ bbkjqh"
  >
    <button
      class="sc-gsnTZi kKvmkM"
      data-testid="button:update"
      type="button"
    />
  </div>
</DocumentFragment>
`;

exports[`<AvailableDays /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-motgx2"
  >
    <div
      aria-pressed="false"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper"
      >
        <span
          class="dot"
        />
        Mon
      </div>
    </div>
    <div
      aria-pressed="false"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper"
      >
        <span
          class="dot"
        />
        Tue
      </div>
    </div>
    <div
      aria-pressed="false"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper"
      >
        <span
          class="dot"
        />
        Wed
      </div>
    </div>
    <div
      aria-pressed="false"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper"
      >
        <span
          class="dot"
        />
        Thu
      </div>
    </div>
    <div
      aria-pressed="false"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper"
      >
        <span
          class="dot"
        />
        Fri
      </div>
    </div>
    <div
      aria-pressed="false"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper"
      >
        <span
          class="dot"
        />
        Sat
      </div>
    </div>
    <div
      aria-pressed="false"
      class="circle-wrapper"
      role="button"
    >
      <div
        class="name-wrapper"
      >
        <span
          class="dot"
        />
        Sun
      </div>
    </div>
  </div>
</DocumentFragment>
`;
